routing_common/transit_graph_data.cpp

```diff
@@ -0,0 +1,564 @@
+  osm::Id const osmId(osmIdNum);
+  auto const it = m_osmIdToFeatureIds.find(osmId);
+  if (it != m_osmIdToFeatureIds.cend())
+  {
+    CHECK_EQUAL(it->second.size(), 1, ("Osm id:", osmId, "(encoded", osmId.EncodedId(),
+        ") from transit graph corresponds to", it->second.size(), "features."
+        "But osm id should be represented be one feature."));
+    id.SetFeatureId(it->second[0]);
+  }
+  id.SetOsmId(osmId.EncodedId());
+}
+
+void DeserializerFromJson::operator()(EdgeFlags & edgeFlags, char const * name)
+{
+  bool transfer = false;
+  (*this)(transfer, name);
+  // Note. Only |transfer| field of |edgeFlags| may be set at this point because the
+  // other fields of |edgeFlags| are unknown.
+  edgeFlags.SetFlags(0);
+  edgeFlags.m_transfer = transfer;
+}
+
+void DeserializerFromJson::operator()(StopIdRanges & rs, char const * name)
+{
+  vector<StopId> stopIds;
+  (*this)(stopIds, name);
+  rs = StopIdRanges({stopIds});
+}
+
+// GraphData --------------------------------------------------------------------------------------
+void GraphData::DeserializeFromJson(my::Json const & root, OsmIdToFeatureIdsMap const & mapping)
+{
+  DeserializerFromJson deserializer(root.get(), mapping);
+  Visit(deserializer);
+
+  // Removes equivalent edges from |m_edges|. If there are several equivalent edges only
+  // the most lightweight edge is left.
+  // Note. It's possible that two stops are connected with the same line several times
+  // in the same direction. It happens in Oslo metro (T-banen):
+  // https://en.wikipedia.org/wiki/Oslo_Metro#/media/File:Oslo_Metro_Map.svg branch 5.
+  my::SortUnique(m_edges,
+                 [](Edge const & e1, Edge const & e2) {
+                   if (e1 != e2)
+                     return e1 < e2;
+                   return e1.GetWeight() < e2.GetWeight();
+                 },
+                 [](Edge const & e1, Edge const & e2) { return e1 == e2; });
+}
+
+void GraphData::Serialize(Writer & writer)
+{
+  auto const startOffset = writer.Pos();
+  Serializer<Writer> serializer(writer);
+  FixedSizeSerializer<Writer> numberSerializer(writer);
+  m_header.Reset();
+  numberSerializer(m_header);
+  m_header.m_stopsOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_stops);
+  m_header.m_gatesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_gates);
+  m_header.m_edgesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_edges);
+  m_header.m_transfersOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_transfers);
+  m_header.m_linesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_lines);
+  m_header.m_shapesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_shapes);
+  m_header.m_networksOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
```

Comment:
I guess it's better to keep ```m_header.m_networksOffset``` closer to ```serializer(m_networks)```, than to ```serializer(m_shapes)```.

Reply:
Done. I did if for all transit types.